File: hubspot/custom-code/customDeleteTasks.js
const dotenv = require("dotenv");

// Utiliza o arquivo .env para obter o accessToken (Importante: nunca mandar o accessToken para o repositório!)
dotenv.config();

// Ao mandar pro custom code do fluxo, copiar a partir desta linha até o comentário [FINAL]
const axios = require("axios");
const secret = process.env.HS_APP;

async function getAssociations(objectType, objectId, toObjectType) {
  try {
    const url = `https://api.hubapi.com/crm/v4/objects/${objectType}/${objectId}/associations/${toObjectType}?limit=500`;
    const headers = { authorization: `Bearer ${secret}` };

    const associations = await axios({ url, method: "get", headers });

    return associations;
  } catch (err) {
    console.error(
      "Erro ao obter associações:",
      err.response ? err.response.data : err.message
    );
    throw err;
  }
}

async function getTask(taskId) {
  try {
    const url = `https://api.hubapi.com/crm/v3/objects/tasks/${taskId}?properties=hs_object_source&properties=hs_task_status&archived=false`;
    const headers = { authorization: `Bearer ${secret}` };

    const task = await axios({ url, method: "get", headers });

    return task;
  } catch (err) {
    console.error(
      "Erro ao obter tarefa:",
      err.response ? err.response.data : err.message
    );
    throw err;
  }
}

async function deleteTask(taskToDelete) {
  try {
    const url = `https://api.hubapi.com/crm/v3/objects/tasks/${taskToDelete}`;
    const headers = { authorization: `Bearer ${secret}` };

    const taskDeleted = await axios({ url, method: "delete", headers });

    return taskDeleted.data;
  } catch (err) {
    console.error(
      "Erro ao excluir tarefa:",
      err.response ? err.response.data : err.message
    );
    throw err;
  }
}

exports.main = async (event, callback) => {
  const { objectId } = event.inputFields;

  try {
    const associations = await getAssociations("leads", objectId, "task");
    const {
      data: { results },
    } = associations;

    for (const result of results) {
      const taskId = result.toObjectId;
      const task = await getTask(taskId);

      const objectSource = task.data.properties.hs_object_source;
      const taskStatus = task.data.properties.hs_task_status;

      if (
        objectSource === "AUTOMATION_PLATFORM" &&
        taskStatus === "NOT_STARTED"
      ) {
        console.log("TASK PARA DELETAR", task.data);
        await deleteTask(taskId);
      }
    }

    callback({
      outputFields: { objectId },
    });
  } catch (error) {
    console.error(
      "Ocorreu um erro inesperado:",
      err.response ? err.response.data : err.message
    );
    throw error;
  }
};

// [FINAL]

exports.main(
  {
    inputFields: {},
    object: { objectId: 666244651 },
  },
  console.log
);

File: hubspot/custom-code/python/customDeleteTasks.py
import requests
import os
from hubspot import HubSpot

hs_app = os.environ.get('hs_app')

# FUNCTIONS
# request associations
def REQUEST_ASSOCIATIONS(object_type, object_id, to_object_type):
  url = f'https://api.hubapi.com/crm/v4/objects/{object_type}/{object_id}/associations/{to_object_type}?limit=500'
  headers = {'authorization': f'Bearer {hs_app}'}

  req_association = requests.get(url, headers=headers)
  return req_association

# request a task
def REQUEST_A_TASK(task_id, hs_app):
  url = f'https://api.hubapi.com/crm/v3/objects/tasks/{task_id}?properties=hs_object_source&properties=hs_task_status&archived=false'
  headers = {'authorization': f'Bearer {hs_app}'}

  req_task = requests.get(url, headers=headers)
  return req_task

# delete a task
def DELETE_A_TASK(task_to_delete, hs_app):
  url = f'https://api.hubapi.com/crm/v3/objects/tasks/{task_to_delete}'
  headers = {'authorization': f'Bearer {hs_app}'}

  requests.delete(url, headers=headers)

# EXECUTION

# Hubspot default function
def main(event):
  object_id = str(event["inputFields"]["object_id"])

  object_type, to_object_type = 'leads', 'task'
  req_association = REQUEST_ASSOCIATIONS(object_type, object_id, to_object_type)

  for i in req_association.json()['results']:
    task_id = str(i['toObjectId'])
    req_task = REQUEST_A_TASK(task_id, hs_app)

    # select tasks with below parameters
    hs_object_source = req_task.json()['properties']['hs_object_source'] == 'AUTOMATION_PLATFORM'
    hs_task_status = req_task.json()['properties']['hs_task_status'] == 'NOT_STARTED'

    if hs_object_source and hs_task_status:
      task_to_delete = task_id
      DELETE_A_TASK(task_to_delete, hs_app)

  return {
  "outputFields": {
    "object_id": object_id
    }
  }